File: backend/app/services/cosmetics.py
"""Cosmetics — the koku desire sink (engagement v2). Server-authoritative ownership.

Cosmetics are pure sinks: they cost koku and grant no gameplay advantage, so the catalog can be
priced aspirationally without unbalancing the economy. The SERVER owns the catalog (price + gate) so
a patched client can't grant itself a skin or pay the wrong price; the mobile app keeps a parallel
catalog only for rendering (ids must match).

Slots hold one equipped item each (e.g. the "sensei" slot = which mascot skin shows). The starter
item of every slot is implicit: always owned, equipped by default, never stored on the row — so a
fresh account already wears something (endowment effect) with no migration backfill.

Gates:
- "buy"     purchasable with koku (price > 0).
- "starter" the free default of its slot; always owned, can't be bought (price 0).
- "achievement" granted server-side on a qualifying event, never for sale (price 0). None ship in
                the first slice — the field exists so the catalog shape is stable.
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from fastapi import HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from ..db.models import User

# Seasonal windows (northern-hemisphere months) — a scarcity/urgency mechanic: a seasonal cosmetic
# can only be bought while its season is active, giving a recurring reason to return and spend.
SEASON_MONTHS: dict[str, set[int]] = {
    "spring": {3, 4, 5},
    "summer": {6, 7, 8},
    "autumn": {9, 10, 11},
    "winter": {12, 1, 2},
}


def is_season_active(season: str | None, now: datetime | None = None) -> bool:
    """True when `season` is None (always available) or the current month falls in its window."""
    if not season:
        return True
    months = SEASON_MONTHS.get(season)
    if not months:
        return True  # unknown season tag → don't lock out (fail open)
    return (now or datetime.now(timezone.utc)).month in months

# id -> {slot, gate, price, season?}. Single source of truth for price + gate.
CATALOG: dict[str, dict[str, Any]] = {
    # Sensei skins (flagship slot — the mascot is on every screen, so highest perceived value).
    "sensei_classic": {"slot": "sensei", "gate": "starter", "price": 0},
    "sensei_sage": {"slot": "sensei", "gate": "buy", "price": 200},
    "sensei_ronin": {"slot": "sensei", "gate": "buy", "price": 250},
    "sensei_sakura": {"slot": "sensei", "gate": "buy", "price": 300, "season": "spring"},
    # Belt-knot styles (shows on the belt icon in the top bar, lists, progress — high daily exposure).
    "knot_classic": {"slot": "knot", "gate": "starter", "price": 0},
    "knot_gold": {"slot": "knot", "gate": "buy", "price": 150},
    "knot_jade": {"slot": "knot", "gate": "buy", "price": 180},
    "knot_tassel": {"slot": "knot", "gate": "buy", "price": 220},
}

_UNKNOWN = "Unknown cosmetic."
_NOT_FOR_SALE = "This cosmetic can't be bought."
_NOT_ENOUGH = "Not enough koku."
_NOT_OWNED = "You don't own this cosmetic."
_OUT_OF_SEASON = "This cosmetic is out of season."


def starter_ids() -> set[str]:
    """The implicit always-owned default of every slot."""
    return {cid for cid, c in CATALOG.items() if c["gate"] == "starter"}


def _starter_for(slot: str) -> str | None:
    for cid, c in CATALOG.items():
        if c["slot"] == slot and c["gate"] == "starter":
            return cid
    return None


def owned_ids(user: User) -> list[str]:
    """Everything the user can equip: implicit starters + purchased/granted ids (deduped, stable)."""
    seen: dict[str, None] = {}
    for cid in list(starter_ids()) + list(user.cosmetics or []):
        if cid in CATALOG:
            seen.setdefault(cid, None)
    return list(seen.keys())


def equipped_resolved(user: User) -> dict[str, str]:
    """Equipped id per slot, falling back to each slot's starter when unset/invalid."""
    out: dict[str, str] = {}
    slots = {c["slot"] for c in CATALOG.values()}
    chosen = user.equipped or {}
    owned = set(owned_ids(user))
    for slot in slots:
        pick = chosen.get(slot)
        if not (isinstance(pick, str) and pick in owned and CATALOG.get(pick, {}).get("slot") == slot):
            pick = _starter_for(slot)
        if pick:
            out[slot] = pick
    return out


def can_buy(user: User, cosmetic_id: str) -> tuple[bool, str]:
    """Pure precheck (no DB write): may this user buy this item now? (ok, reason)."""
    item = CATALOG.get(cosmetic_id)
    if item is None:
        return False, _UNKNOWN
    if item["gate"] != "buy":
        return False, _NOT_FOR_SALE
    if cosmetic_id in owned_ids(user):
        return False, "Already owned."
    if not is_season_active(item.get("season")):
        return False, _OUT_OF_SEASON
    if user.coins < item["price"]:
        return False, _NOT_ENOUGH
    return True, ""


async def buy(user: User, db: AsyncSession, cosmetic_id: str) -> User:
    """Purchase a cosmetic: validate gate/season/price, debit koku, grant ownership — atomically.

    Row-locks the account (SELECT … FOR UPDATE) so concurrent buys serialize. Without the lock two
    buys of the SAME item could both pass the ownership check and double-charge, and two DIFFERENT
    buys could clobber each other's `cosmetics` append (a read-modify-write on the JSONB array).
    409 when too poor (re-checked under the lock)."""
    item = CATALOG.get(cosmetic_id)
    if item is None:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=_UNKNOWN)
    if item["gate"] != "buy":
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=_NOT_FOR_SALE)
    # Lock the row for the read-modify-write below (ownership re-check + debit + array append).
    await db.refresh(user, with_for_update=True)
    if cosmetic_id in owned_ids(user):
        await db.rollback()  # idempotent: already owned → release the lock, charge nothing
        return user
    if not is_season_active(item.get("season")):
        await db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=_OUT_OF_SEASON)
    price = item["price"]
    if user.coins < price:
        await db.rollback()
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=_NOT_ENOUGH)
    user.coins -= price
    user.cosmetics = list(user.cosmetics or []) + [cosmetic_id]
    await db.commit()
    await db.refresh(user)
    return user


async def equip(user: User, db: AsyncSession, cosmetic_id: str) -> User:
    """Equip an owned cosmetic into its slot. 400 if unknown; 409 if not owned."""
    item = CATALOG.get(cosmetic_id)
    if item is None:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=_UNKNOWN)
    if cosmetic_id not in owned_ids(user):
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=_NOT_OWNED)
    equipped = dict(user.equipped or {})
    equipped[item["slot"]] = cosmetic_id
    user.equipped = equipped
    await db.commit()
    await db.refresh(user)
    return user

File: mobile/store/cosmetics.ts
// Cosmetics — the client mirror of the server catalog (engagement v2: koku desire sink).
//
// Ownership is SERVER-authoritative (services/cosmetics.py); this module is pure and exists for two
// jobs: (1) render — it holds each cosmetic's display name + visual spec; (2) UI gating — cheap
// pure prechecks (owned? affordable? buyable?) that mirror the server rules so the Wardrobe can grey
// out / label items without a round-trip. The server still enforces price + gate on every buy.
//
// Ids and prices MUST match backend services/cosmetics.py CATALOG.

export type Slot = "sensei";
export type Gate = "starter" | "buy" | "achievement";

// Visual override for the Sensei mascot skin. Empty = the default (classic) look.
export type SenseiVisual = {
  hair?: string;
  skinTone?: string;
  skinEdge?: string;
  accessory?: "beard" | "scar" | "sakura";
};

export type CosmeticDef = {
  id: string;
  slot: Slot;
  gate: Gate;
  price: number;
  season?: string;
  name: { en: string; ru: string };
  blurb: { en: string; ru: string };
  visual: SenseiVisual;
};

export const CATALOG: Record<string, CosmeticDef> = {
  sensei_classic: {
    id: "sensei_classic",
    slot: "sensei",
    gate: "starter",
    price: 0,
    name: { en: "Classic Sensei", ru: "Классический Сэнсэй" },
    blurb: { en: "The master you started with.", ru: "Мастер, с которого всё началось." },
    visual: {},
  },
  sensei_sage: {
    id: "sensei_sage",
    slot: "sensei",
    gate: "buy",
    price: 200,
    name: { en: "Sage", ru: "Мудрец" },
    blurb: { en: "Silver hair, calm eyes.", ru: "Седина и спокойный взгляд." },
    visual: { hair: "#D9D9D2", accessory: "beard" },
  },
  sensei_ronin: {
    id: "sensei_ronin",
    slot: "sensei",
    gate: "buy",
    price: 250,
    name: { en: "Ronin", ru: "Ронин" },
    blurb: { en: "A wanderer with a past.", ru: "Странник с прошлым." },
    visual: { hair: "#16161B", skinTone: "#E9C49E", skinEdge: "#D2A877", accessory: "scar" },
  },
  sensei_sakura: {
    id: "sensei_sakura",
    slot: "sensei",
    gate: "buy",
    price: 300,
    season: "spring",
    name: { en: "Sakura", ru: "Сакура" },
    blurb: { en: "Petals of the spring festival.", ru: "Лепестки весеннего фестиваля." },
    visual: { hair: "#3A2A30", accessory: "sakura" },
  },
};

export const SLOTS: Slot[] = ["sensei"];

export function defOf(id: string): CosmeticDef | undefined {
  return CATALOG[id];
}

export function catalogForSlot(slot: Slot): CosmeticDef[] {
  return Object.values(CATALOG).filter((c) => c.slot === slot);
}

export function starterFor(slot: Slot): string {
  const s = Object.values(CATALOG).find((c) => c.slot === slot && c.gate === "starter");
  return s ? s.id : "";
}

export function isOwned(owned: string[] | undefined, id: string): boolean {
  const def = CATALOG[id];
  if (!def) return false;
  if (def.gate === "starter") return true; // starters are always owned
  return !!owned && owned.includes(id);
}

export function canAfford(coins: number, id: string): boolean {
  const def = CATALOG[id];
  return !!def && coins >= def.price;
}

export type BuyCheck = { ok: boolean; reason: "buyable" | "owned" | "not_for_sale" | "too_poor" | "unknown" };

/** Pure UI precheck mirroring the server's buy rules (the server still enforces). */
export function buyCheck(coins: number, owned: string[] | undefined, id: string): BuyCheck {
  const def = CATALOG[id];
  if (!def) return { ok: false, reason: "unknown" };
  if (def.gate !== "buy") return { ok: false, reason: "not_for_sale" };
  if (isOwned(owned, id)) return { ok: false, reason: "owned" };
  if (coins < def.price) return { ok: false, reason: "too_poor" };
  return { ok: true, reason: "buyable" };
}

/** The visual spec for the equipped Sensei skin, falling back to the classic default. */
export function senseiVisual(equipped: Record<string, string> | undefined): SenseiVisual {
  const id = (equipped && equipped.sensei) || starterFor("sensei");
  return CATALOG[id]?.visual ?? {};
}
